internal/pubsub/example_test.go

```diff
@@ -101,3 +103,147 @@ func Example_sendReceiveMultipleMessages() {
 	// b
 	// c
 }
+
+func Example_receiveWithInvertedWorkerPool() {
+	// Open a topic and corresponding subscription.
+	ctx := context.Background()
+	b := mempubsub.NewBroker([]string{"myTopic"})
+	t := mempubsub.OpenTopic(b, "myTopic")
+	defer t.Close()
+	s := mempubsub.OpenSubscription(b, "myTopic", time.Second)
+	defer s.Close()
+
+	// Send a bunch of messages to the topic.
+	const nMessages = 100
+	for n := 0; n < nMessages; n++ {
+		m := &pubsub.Message{
+			Body: []byte(fmt.Sprintf("message %d", n)),
+		}
+		if err := t.Send(ctx, m); err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	// In order to make our test exit, we keep track of how many messages were
+	// processed and cancel the receiveCtx when we've processed them all.
+	var nProcessed int32
+	receiveCtx, cancel := context.WithCancel(ctx)
+	go func() {
+		for {
+			if atomic.LoadInt32(&nProcessed) == nMessages {
+				cancel()
+				break
+			}
+		}
+	}()
+
+	// Process messages using an inverted worker pool, as described here:
+	// https://www.youtube.com/watch?v=5zXAHh5tJqQ&t=26m58s
+	// It uses a buffered channel, sem, as a semaphore to manage the maximum
+	// number of workers.
+	const poolSize = 10
+	sem := make(chan struct{}, poolSize)
+	for {
```

Comment:
Why not just
```
for atomic.LoadInt32(&nProcessed) < nMessages {
```

Reply:
I think this is more realistic; normal users will likely have an infinite loop with a `break` rather than an explicit condition. Also, I find that misleading, as you might think it would work without the cancellation of the ctx, but it doesn't -- you can enter the loop when `nMessages` is small, call `Receive`, and block forever as `nMessages` is incremented.